---
title: "PILRB1 - Differential Expression Analysis"
output: html_notebook
---

#
#### R Version
Built with `r getRversion()`


#### Load data
```{r, echo=FALSE}
#setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
load("../data/rnaseq/processed/Processed_RNAseq.Rdata")
```

#### Set color scheme and breaks
```{r, results="hide", fig.height = 1.5, fig.width = 1.5}
library(RColorBrewer)

# Cluster palette
clust.pal <- brewer.pal(8, "Spectral")

# Heatmap colors
jet <- c("#7F0000","red","#FF7F00","yellow","white","cyan", "#007FFF", "blue","#00007F")
colorsZ <- colorRampPalette(rev(jet))(1000)

mb.col <- c("#023858","#3690c0","#e0f3f8","#fff7bc","#fec44f","#ec7014","#fc4e2a","#bd0026","#800026")
colorsCPM <- colorRampPalette(mb.col)(n=1000)

# Generate the breaks for the color scale
col.breaksZ <- seq(-3, 3, length.out = 1001)
col.breaksCPM <- seq(0,16,length = 1001)

# Look at color assignments
pie(x = rep(1,length(jet)), col = rev(jet), main = "Z-score Key", labels = seq(-3,3,0.75))
pie(x = rep(1,length(mb.col)), col = mb.col, main = "CPM Key", labels = seq(0,16,2))

```

## Differential Expression Analysis

### #Make the model matrix for analysis
```{r}
library(edgeR)

# Set up ANOVA-like difference
meta$Group <- as.factor(meta$Group)

mod <- model.matrix(~0+meta$Group)
colnames(mod) <- gsub("meta\\$", "", colnames(mod))
colnames(mod) <- gsub("Group", "", colnames(mod))
mod

# Estimate dispersion
gene.dge.filt <- estimateDisp(gene.dge.filt, mod)

# Fit
fit <- glmQLFit(gene.dge.filt, mod)
colnames(fit)

# Set desired individual contrasts
conMatrix <- makeContrasts(
  WT_v_KO = (P28_KO + M12_KO)/2 - (P28_WT + M12_WT)/2,
  P28_v_M12 = (M12_KO + M12_WT)/2 - (P28_KO + P28_WT)/2,
  P28_WTvKO = P28_KO - P28_WT,
  M12_WTvKO = M12_KO - M12_WT,
  levels=mod)
print("Matrix of examined contrasts")
conMatrix


```

#
####Perform differential expression 
```{r}

# Set chromosome list
chrList <- c(1:22,"Y","X")

# Set thresholds
library(dplyr)
fc <- 1
fdr <- 0.05

   
## Run Quasi-likelihood F Tests for all comparisons
qlf_WT_v_KO <- glmQLFTest(fit, contrast = conMatrix[,"WT_v_KO"])
qlf_P28_v_M12 <- glmQLFTest(fit, contrast = conMatrix[,"P28_v_M12"])
qlf_P28_WTvKO <- glmQLFTest(fit, contrast = conMatrix[,"P28_WTvKO"])
qlf_M12_WTvKO <- glmQLFTest(fit, contrast = conMatrix[,"M12_WTvKO"])

## Store all genes 
DE <- c()
DE$WT_v_KO <- data.frame(topTags(qlf_WT_v_KO, n = Inf, p.value = 1, adjust.method = "BH", sort.by = "logFC")) %>% tibble::rownames_to_column("ensembl_ID")
DE$P28_v_M12 <- data.frame(topTags(qlf_P28_v_M12, n = Inf, p.value = 1, adjust.method = "BH", sort.by = "logFC")) %>% tibble::rownames_to_column("ensembl_ID")
DE$P28_WTvKO <- data.frame(topTags(qlf_P28_WTvKO, n = Inf, p.value = 1, adjust.method = "BH", sort.by = "logFC")) %>% tibble::rownames_to_column("ensembl_ID")
DE$M12_WTvKO <- data.frame(topTags(qlf_M12_WTvKO, n = Inf, p.value = 1, adjust.method = "BH", sort.by = "logFC")) %>% tibble::rownames_to_column("ensembl_ID")

## Store sig genes
DE$WT_v_KO_sig <-data.frame(topTags(qlf_WT_v_KO, n = Inf, p.value = fdr, adjust.method = "BH", sort.by = "logFC")) %>% filter(abs(logFC) > fc) %>% filter(chromosome_name %in% chrList) %>% tibble::rownames_to_column("ensembl_ID")
DE$P28_v_M12_sig <-data.frame(topTags(qlf_P28_v_M12, n = Inf, p.value = fdr, adjust.method = "BH", sort.by = "logFC")) %>% filter(abs(logFC) > fc) %>% filter(chromosome_name %in% chrList) %>% tibble::rownames_to_column("ensembl_ID")
DE$P28_WTvKO_sig <-data.frame(topTags(qlf_P28_WTvKO, n = Inf, p.value = fdr, adjust.method = "BH", sort.by = "logFC")) %>% filter(abs(logFC) > fc) %>% filter(chromosome_name %in% chrList) %>% tibble::rownames_to_column("ensembl_ID")
DE$M12_WTvKO_sig <-data.frame(topTags(qlf_M12_WTvKO, n = Inf, p.value = fdr, adjust.method = "BH", sort.by = "logFC")) %>% filter(abs(logFC) > fc) %>% filter(chromosome_name %in% chrList) %>% tibble::rownames_to_column("ensembl_ID")


## Sig genes merged with CPM data
cpm_table <- as.data.frame(gene.dge.filt$cpm)  %>% tibble::rownames_to_column("ensembl_ID")
DE$WT_v_KO_sig_cpm <- DE$WT_v_KO_sig %>% left_join(cpm_table)
DE$P28_v_M12_sig_cpm <- DE$P28_v_M12_sig %>% left_join(cpm_table)
DE$P28_WTvKO_sig_cpm <- DE$P28_WTvKO_sig %>% left_join(cpm_table)
DE$M12_WTvKO_sig_cpm <- DE$M12_WTvKO_sig %>% left_join(cpm_table)

#Export Results
write.csv(DE$WT_v_KO_sig_cpm, "../data/rnaseq/processed/DEG_WT_v_KO.csv", row.names = TRUE)
write.csv(DE$P28_v_M12_sig_cpm, "../data/rnaseq/processed/DEG_P28_v_M12_sig.csv", row.names = TRUE)
write.csv(DE$P28_WTvKO_sig_cpm, "../data/rnaseq/processed/DEG_P28_WTvKO.csv", row.names = TRUE)
write.csv(DE$M12_WTvKO_sig_cpm, "../data/rnaseq/processed/DEG_M12_WTvKO.csv", row.names = TRUE)

```

#
####Average the data and get Z-score
```{r}
source("../data/rnaseq/src/MBave.R")


# Get the average log2 CPM expression per group for plotting
gene.dge$ave <- MBave(dat = gene.dge$lcpm, grp = grp)
gene.dge$cpm_ave <- MBave(dat = gene.dge$cpm, grp = grp)
```


```{r}
# Get the averaged Z-score for the DE sig genes
gene.dge$aveZ <- t(scale(t(gene.dge$ave[DE$P28_WTvKO_sig$ensembl_ID,]), center = T, scale = T))
```

```{r}
save.image(file = "../data/rnaseq/processed/Processed_RNAseq_w_CPM.Rdata")
```

